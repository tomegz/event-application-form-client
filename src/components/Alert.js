import React from "react";
import PropTypes from "prop-types";

class Alert extends React.Component {
  componentDidMount() {
    this.timer = setTimeout(
      this.props.onClose,
      this.props.timeout
    );
  }
  componentWillUnmount() {
    clearTimeout(this.timer);
  }
  render() {
    return (
      <div className={`alert alert-${this.props.message.type}`}>
        {this.props.message.text}
        <button 
          className={`btn-close alert-${this.props.message.type}`}
          onClick={this.props.onClose}
        >
          &times;
        </button>
      </div>
    );
  }
}

Alert.propTypes = {
  onClose: PropTypes.func,
  timeout: PropTypes.number,
  message: PropTypes.object.isRequired
};

Alert.defaultProps = {
  timeout: 3000
};

export default Alert;