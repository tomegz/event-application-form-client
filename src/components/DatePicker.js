import React from "react";
import Calendar from "react-calendar";

class DatePicker extends React.Component {
  constructor(props) {
    super(props);
    this.handleChange = this.handleChange.bind(this);
  }
  handleChange(date) {
    this.props.setDate(date);
  }
  render() {
    const tomorrow = new Date(new Date().getTime() + 24 * 60 * 60 * 1000);
    return (
      <div className="react-calendar-wrapper">
        <Calendar locale="en-EN" onChange={this.handleChange} minDate={tomorrow} />
      </div>
    );
  }
} 

export default DatePicker;