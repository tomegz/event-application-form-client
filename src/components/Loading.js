import React from 'react';

class Loading extends React.Component {
  render() {
    return (
      <div className="loader-wrapper">
        <div className="lds-ellipsis"><div></div><div></div><div></div><div></div></div>
      </div>
    );
  }
}

export default Loading;