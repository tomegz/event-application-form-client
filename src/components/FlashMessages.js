import React from "react";
import Alert from "./Alert";
import { connect } from "react-redux";
import { removeMessage } from "../actions/actionCreators";

class FlashMessages extends React.Component {
  render() {
    return (
      <div>
        {this.props.messages.map((message, i) => {
          return (
            <Alert 
              key={message.id} 
              message={message} 
              onClose={() => this.props.removeMessage(message.id)} 
            />
          );
        })}
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    messages: state.messages
  }
}

export default connect(mapStateToProps, { removeMessage })(FlashMessages);