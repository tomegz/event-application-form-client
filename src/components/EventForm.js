import React from "react";
import DatePicker from "./DatePicker";
import Loading from "./Loading";
import postEvent from "../utils/api";
import shortid from "shortid";

import { connect } from "react-redux";
import { addMessage, updateField, resetFields } from "../actions/actionCreators";

export class EventForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      calendarVisible: false,
      loading: false
    }
    this.showCalendar = this.showCalendar.bind(this);
    this.hideCalendar = this.hideCalendar.bind(this);
    this.setDate = this.setDate.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.clearForm = this.clearForm.bind(this);
  }
  showCalendar() {
    this.setState(() => {
      return {
        calendarVisible: true
      }
    });
  }
  hideCalendar() {
    this.setState(() => {
      return {
        calendarVisible: false
      }
    });
  }
  clearForm() {
    this.setState(() => {
      return {
        loading: false
      };
    })
    this.props.resetFields();
  }
  setDate(date) {
    const timestamp = date.getTime();
    const year = date.getFullYear();
    const month = date.getMonth() + 1;
    const day = date.getDate();
    const eventDate = {
      timestamp,
      year,
      month,
      day
    }
    this.props.updateField("eventDate", eventDate);
    this.setState(() => {
      return {
        calendarVisible: false
      }
    });
  }
  handleChange(e) {
    const { name, value } = e.target;
    this.props.updateField(name, value);
  }
  handleSubmit(e) {
    const { firstName, lastName, email } = this.props.fields;
    const eventDate = this.props.fields.eventDate.timestamp;
    const data = { firstName, lastName, email, eventDate };
    this.setState(() => {
      return {
        loading: true
      }
    });
    postEvent('/events', data)
      .then(res => {
        res.messages.forEach(message => this.props.addMessage({ ...message, id: shortid.generate() }));
        this.clearForm();
      }) 
      .catch(error => console.error(error));
    e.preventDefault();
  }
  render() {
    const { firstName, lastName, email } = this.props.fields;
    const { day, month, year } = this.props.fields.eventDate;
    const eventDate = day ? `${day}/${month}/${year}` : "";
    return (
      <form className="event-form" onSubmit={this.handleSubmit}>
        <h3>Event application form</h3>
        <label htmlFor="firstName">First Name:</label>
        <input type="text" name="firstName" required placeholder="John" onChange={this.handleChange}  value={firstName}/>
        <label htmlFor="lastName">Last Name:</label>
        <input type="text" name="lastName" required placeholder="Doe" onChange={this.handleChange} value={lastName}/>
        <label htmlFor="email">Email Address:</label>
        <input type="email" name="email" required placeholder="johndoe@example.com" onChange={this.handleChange} value={email}/>
        <label htmlFor="date">Event's date:</label>
        {this.state.calendarVisible 
          ? <DatePicker setDate={this.setDate} hideCalendar={this.hideCalendar} /> 
          : ""}
        <input type="text" name="eventDate" required 
          placeholder="DD/MM/YYYY" 
          value={eventDate}
          onFocus={this.showCalendar} />
        {this.state.loading 
          ? <Loading /> 
          : <button className="event-button" type="submit">Submit</button>}
      </form>
    );
  }
}

const mapStateToProps = state => {
  return {
    fields: state.fields
  }
}

export default connect(mapStateToProps, { addMessage, updateField, resetFields })(EventForm);