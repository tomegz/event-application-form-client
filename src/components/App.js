import React from "react";
import EventForm from "./EventForm";
import FlashMessages from "./FlashMessages";

const App = () => (
  <div className='app'>
    <FlashMessages />
    <EventForm />
  </div>
);

export default App;
