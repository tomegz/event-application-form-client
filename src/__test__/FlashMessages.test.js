import React from "react";
import { FlashMessages } from "../components/FlashMessages";
import renderer from "react-test-renderer";

describe("flash messages - snaphot", () => {
  it("renders correctly", () => {
    const messages = [
      { id: 1, type: "error", text: "Please provide first name" },
      { id: 2, type: "error", text: "Please provide last name" }
    ];
    const tree = renderer.create(<FlashMessages messages={messages}/>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
