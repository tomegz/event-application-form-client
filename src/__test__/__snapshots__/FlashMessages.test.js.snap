// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flash messages - snaphot renders correctly 1`] = `
<div>
  <div
    className="alert alert-error"
  >
    Please provide first name
    <button
      className="btn-close alert-error"
      onClick={[Function]}
    >
      ×
    </button>
  </div>
  <div
    className="alert alert-error"
  >
    Please provide last name
    <button
      className="btn-close alert-error"
      onClick={[Function]}
    >
      ×
    </button>
  </div>
</div>
`;
